quartz/build.ts

```diff
@@ -73,108 +83,122 @@ async function startServing(
 ) {
   const { argv } = ctx
 
-  const ignored = await isGitIgnored()
   const contentMap = new Map<FilePath, ProcessedContent>()
   for (const content of initialContent) {
     const [_tree, vfile] = content
     contentMap.set(vfile.data.filePath!, content)
   }
 
-  const initialSlugs = ctx.allSlugs
-  let lastBuildMs = 0
-  const toRebuild: Set<FilePath> = new Set()
-  const toRemove: Set<FilePath> = new Set()
-  const trackedAssets: Set<FilePath> = new Set()
-  async function rebuild(fp: string, action: "add" | "change" | "delete") {
-    // don't do anything for gitignored files
-    if (ignored(fp)) {
-      return
-    }
+  const buildData: BuildData = {
+    ctx,
+    mut,
+    contentMap,
+    ignored: await isGitIgnored(),
+    initialSlugs: ctx.allSlugs,
+    trackedAssets: new Set<FilePath>(),
+    lastBuildMs: 0,
+  }
 
-    // dont bother rebuilding for non-content files, just track and refresh
-    fp = toPosixPath(fp)
-    const filePath = joinSegments(argv.directory, fp) as FilePath
-    if (path.extname(fp) !== ".md") {
-      if (action === "add" || action === "change") {
-        trackedAssets.add(filePath)
-      } else if (action === "delete") {
-        trackedAssets.delete(filePath)
-      }
-      clientRefresh()
-      return
-    }
+  const watcher = chokidar.watch(".", {
+    persistent: true,
+    cwd: argv.directory,
+    ignoreInitial: true,
+  })
+
+  watcher
+    .on("add", (fp) => rebuild(fp, "add", clientRefresh, buildData))
+    .on("change", (fp) => rebuild(fp, "change", clientRefresh, buildData))
+    .on("unlink", (fp) => rebuild(fp, "delete", clientRefresh, buildData))
+
+  return async () => {
+    await watcher.close()
+  }
+}
+
+async function rebuild(
+  fp: string,
+  action: "add" | "change" | "delete",
+  clientRefresh: () => void,
+  buildData: BuildData, // note: this function mutates buildData
+) {
+  const { ctx, ignored, mut, initialSlugs, contentMap, trackedAssets, lastBuildMs } = buildData
 
+  const { argv } = ctx
+
+  const toRebuild = new Set<FilePath>()
+  const toRemove = new Set<FilePath>()
```

Comment:
these two should probably also go in build data, it is transient but needs to be tracked between rebuilds

(let's also rename `rebuild` to `rebuildFile` or something to also make it more clear it only rebuilds one file, not the whole thing)

Reply:
`toRebuild` and `toRemove` are cleared at the end of the function every time -- could you clarify what is being tracked between rebuilds?

I think `rebuildFile` is a less accurate name. It's true that only the changed file is parsed, but the function currently cleans and re-emits everything in the output directory, which seems to me analogous to a full rebuild with a cache.

Reply:
`toRebuild` and `toRemove` track all the changes that have happened between rebuilds because we do some debouncing there, it would be incorrect to clear it at the start instead

perhaps `rebuildFromEntrypoint` ?

Reply:
Ah, that makes sense, thanks. Move them into `buildData` and also renamed to `rebuildFromEntrypoint`